pkg/datapath/loader/netlink.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	libbpfFixupMsg = "struct bpf_elf_map fixup performed due to size mismatch!"
+)
+
+func replaceQdisc(ifName string) error {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return err
+	}
+	attrs := netlink.QdiscAttrs{
+		LinkIndex: link.Attrs().Index,
+		Handle:    netlink.MakeHandle(0xffff, 0),
+		Parent:    netlink.HANDLE_CLSACT,
+	}
+
+	qdisc := &netlink.GenericQdisc{
+		QdiscAttrs: attrs,
+		QdiscType:  "clsact",
+	}
+
+	// Replacing the qdisc after the first creation will always fail with
+	// the current netlink library due to the issue fixed in this PR:
+	// https://github.com/vishvananda/netlink/pull/382
+	//
+	// FIXME GH-5423 rebase against the latest netlink library
+	if err = netlink.QdiscReplace(qdisc); err != nil {
+		log.WithError(err).Debugf("netlink: Replacing qdisc for %s failed", ifName)
+	} else {
+		log.Debugf("netlink: Replacing qdisc for %s succeeded", ifName)
+	}
+
+	return nil
+}
+
+// replaceDatapath the qdisc and BPF program for a endpoint
+func replaceDatapath(ctx context.Context, ifName string, objPath string, progSec string) error {
+	err := replaceQdisc(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to replace Qdisc for %s: %s", ifName, err)
+	}
+
+	// FIXME: Replace cilium-map-migrate with Golang map migration
+	out, err := exec.CommandContext(ctx, "cilium-map-migrate", "-s", objPath).CombinedOutput()
+	if ctx.Err() == context.DeadlineExceeded {
+		err = fmt.Errorf("Command execution failed: Timeout")
```

Comment:
Having looked into this, I agree, but I'd rather tackle it in a separate PR since there's places around in `daemon/` and potentially other bits that could make use of such a generic helper.

Reply:
PR: #5439

Reply:
Nice, thanks!